pkg/cmd/config.go

```diff
@@ -0,0 +1,141 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    https://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"github.com/spf13/pflag"
+
+	clientset "github.com/mayadata-io/dmaas-operator/pkg/generated/clientset/versioned"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// Config interface provides api to access clientset and namespace
+type Config interface {
+	// BindFlags binds common flags (--kubeconfig, --namespace) to the passed-in FlagSet.
+	BindFlags(flags *pflag.FlagSet)
+	// Client returns dmaas-operator client.
+	Client() (clientset.Interface, error)
+	// KubeClient returns Kubernetes client.
+	KubeClient() (kubernetes.Interface, error)
+	// SetClientQPS sets the Queries Per Second for a client.
+	SetClientQPS(float32)
+	// SetClientBurst sets the Burst for a client.
+	SetClientBurst(int)
+	// GetNamespace return dmaas namespace
+	GetNamespace() string
+}
+
+type config struct {
+	kubeconfig  string
+	namespace   string
+	clientQPS   float32
+	clientBurst int
+	flags       *pflag.FlagSet
+}
+
+// NewConfig return config for clientset
+func NewConfig() Config {
+	c := &config{
+		flags: pflag.NewFlagSet("", pflag.ContinueOnError),
+	}
+
+	c.namespace = "dmaas"
+
+	c.flags.StringVar(&c.kubeconfig,
+		"kubeconfig",
+		"",
+		"Path to the kubeconfig file to use to talk to the Kubernetes apiserver. If unset, try the environment variable KUBECONFIG, as well as in-cluster configuration",
+	)
+	c.flags.StringVarP(&c.namespace,
+		"namespace",
+		"n",
+		c.namespace,
+		"The namespace in which dmaas-operator should operate",
+	)
+
+	return c
+}
+
+// BindFlags binds common flags to the given flagset
+func (c *config) BindFlags(flags *pflag.FlagSet) {
+	flags.AddFlagSet(c.flags)
+}
+
+// ClientConfig return cluster config
+func (c *config) ClientConfig() (*rest.Config, error) {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	loadingRules.ExplicitPath = c.kubeconfig
+	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, nil)
+
+	clientConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "error finding Kubernetes API server config in --kubeconfig, $KUBECONFIG, or in-cluster configuration")
+	}
+
+	if c.clientQPS > 0.0 {
+		clientConfig.QPS = c.clientQPS
+	}
+
+	if c.clientBurst > 0 {
+		clientConfig.Burst = c.clientBurst
+	}
+
+	return clientConfig, nil
+}
+
+// Client returns clientset to access dmaas-operator resources
+func (c *config) Client() (clientset.Interface, error) {
```

Comment:
these functions Client() and KubeClient() doesn't have callers yet..
I see the clientsare generated without any checks in these methods..
but, is my assumption right that these will be called only once or very rarely?
otherwise, it ends up creating many of them

Reply:
yes. This will be called once to fetch clientset and fetched clientset will be assigned to controller. So it will be called once only. I've added code for controller which uses this interface -- https://github.com/mayadata-io/dmaas-operator/pull/4/commits/9e9ec9070990115759a9b27cbda4c1497bebb4a2#diff-a4874ae26fb75dd53cb22801a2441507R153